### PR TITLE
minimal stable diffusion GPU memory usage with accelerate hooks

### PR DESCRIPTION
**Pull request by piEsposito**
Attempts to solve #540 in a more readable, less intrusive, less verbose way than #537.

@patil-suraj I made another attempt to solve this, now in a non-intrusive way, using `accelerate.cpu_offload` to keep on GPU only the parts of the models that are being used on operations, keeping the GPU memory footprint as little as `800 mb`.

Would you be so gentle as to take a look and tell me if I'm going on the right direction?

Thanks!

Changes:
- If pipeline would return `torch.device("meta")` on the `.device` getter, return `torch.device("cpu")`, otherwise we won't be able to move tensors to `self.device` if using accelerate
- Created `StableDiffusionPipeline.cuda_with_minimal_gpu_usage` method to `acclerate.cpu_offload` all models to reduce GPU memory footprint.
- Made PR tests depend on `accelerate` from master, as per @sgugger suggestion.

We need to install `accelerate` from master to use it, as this PR depends on https://github.com/huggingface/accelerate/pull/768.

### CODE REVIEW

**Comment by hafriedlander** [comment]
Just for reference, this is how I solved the same problem - https://github.com/hafriedlander/stable-diffusion-grpcserver/blob/b75deaa743c77415f277fedd37494a661a0cbaf2/sdgrpcserver/pipeline/unified_pipeline.py#L364. 

This gets used as a baseclass for a pipeline rather than DiffusionPipeline, and moves one model at a time onto the GPU. The advantage compared to this is that all the models run on the GPU, the disadvantage is that it's dependent on garbage collection to free the GPU memory.

**Comment by piEsposito** [comment]
@hafriedlander thank you for the suggestion. I found out that we can [do that with accelerate](https://huggingface.co/docs/accelerate/v0.13.1/en/package_reference/big_modeling#accelerate.hooks.attach_execution_device_hook): if we add this hook to each model, it will offload to GPU onnly when performing inference, keeping the memory usage minimal.

I tried to avoid losing time with I/O and thus did not added everything to be loaded to GPU. Nevertheless, I agree with you this is a possible add-on to the PR.

**Comment by patrickvonplaten** [comment]
Hey @piEsposito, 

Generally, I think I'm fine with this PR! Just one question: In our experiments it's better and faster to run the model in pure "fp16" and to **not** use autocast at all - see: https://github.com/huggingface/diffusers/pull/371 

Should we maybe try this here as well? E.g. just removing the `autocast` call?

**Comment by patrickvonplaten** [comment]
Otherwise happy to add this to the pipeline :-)

**Comment by piEsposito** [comment]
@patrickvonplaten thank you for your feedback.

About your concern: I think we can't run the model in pure fp16 because some layers (e.g. batchnorm) are implemented only for fp32. Because of that we can oly keep the unet on fp16 (as it will run on the GPU), and the other models need to be in fp32 to be able to run on CPU.

That way, the autocast works as a way to put the tensors in the right precision without having to intrude on the pipeline code.

Please let me know if there is a way to keep them on fp16 o CPU and be able to run those layers, if that's the case I can add it to the PR.

Thanks!

**Comment by piEsposito** [comment]
@patrickvonplaten an alternative would be to use [`accelerate.cpu_offload`](https://github.com/huggingface/accelerate/blob/8b16276a41857e25f8e0d0cfe45eb69c60134add/src/accelerate/big_modeling.py#L112-L155) (so that we keep everything on CPU and load to GPU only when running the model). This alternative has the following tradeoff:

- For one side, we can run the model in full `fp16`, for other we have a lot of I/O between CPU and GPU when offloading the model in and out of the devices for the forward pass
- We would have to tweak the [`DiffusionPipeline.device`](https://github.com/huggingface/diffusers/blob/cca59ce3a289d8837d81747389f997d6a1564d1d/src/diffusers/pipeline_utils.py#L202-L212) getter to, in case it returns a `torch.device("meta")`, return `torch.device("cpu")`. Otherwise, when we use `accelerate.cpu_offload`, it will move the models to a meta device, then the `device` attribute of the pipeline will start returning `meta` and it will break when we try to move a tensor to the pipeline `device` attribute.

Both ways work, but if we go for `accelerate.cpu_offload`ing all the model, I would just like an ok of yours before changing the pipeline abstraction.

Which alterantive do you think is best?

**Comment by piEsposito** [comment]
Update: I've opened https://github.com/huggingface/accelerate/pull/768 to solve the issue with `accelerate.cpu_offload` and the execution devices.

If the PR on accelerate is merge, we can change this PR's

https://github.com/huggingface/diffusers/blob/0b601e63f83e1a5cd9c2dc2dc70ac0e9bdf404ff/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L122-L137

to

```python
def cuda_with_minimal_gpu_usage(self):
      if is_accelerate_available():
          from accelerate import cpu_offload
      else:
          raise ImportError("Please install accelerate via `pip install accelerate`")

      device = torch.device("cuda")
      self.enable_attention_slicing(1)

      for cpu_offloaded_model in [self.unet, self.text_encoder, self.vae, self.safety_checker]:
          cpu_offload(cpu_offloaded_model, device)
```

And reduce memory usage even further thanks to CPU offload.

**Comment by piEsposito** [comment]
@patrickvonplaten I will wait for `accelerate` to have a release with this commit (https://github.com/huggingface/accelerate/commit/5e8ab12c3d328347aa565105ddce86e35950ab32 from https://github.com/huggingface/accelerate/pull/768) merged, so that I can just do `cpu_offload` and the PR as clarer.

Would that be ok?

**Comment by piEsposito** [comment]
@sgugger, can I ask for a patch release of accelerate with (https://github.com/huggingface/accelerate/commit/5e8ab12c3d328347aa565105ddce86e35950ab32 from https://github.com/huggingface/accelerate/pull/768) merged? It would unblock this PR.

**Comment by sgugger** [comment]
Not sure we will do a patch release for that, it's not a regression introduced by anything, just a bug fix. The big model inference in Accelerate is getting improved every day, so Diffusers should test again the main branch of Accelerate for now (that's what we do in Transformers).

**Comment by patrickvonplaten** [comment]
Hey @piEsposito,

Cool to see that the change has been merged to `accelerate` - If you want feel free to go ahead with making this PR simpler and then we'll just update the requirement to `accelerate>=0.14.0` since it's currently on `0.14.0.dev0` meaning the next minor version will be `0.14.0` (see: https://github.com/huggingface/accelerate/blob/87a7e0783f7b428fafa72a9e64d36625ee22888f/src/accelerate/__init__.py#L5) 

Then we can just ask people to install accelerate from github until `0.14.0 ` :-):

```
pip install git+https://github.com/huggingface/accelerate.git
```

Does this sound good to you? :-)

**Comment by piEsposito** [comment]
@patrickvonplaten this is perfect, I'm on it.

**Comment by patrickvonplaten** [comment]
Ah I see regarding the tests, let me and @anton-l take care of it :-) Don't worry about the failing test we'll make a change to the testing files

File: .github/workflows/pr_tests.yml
Comment (patrickvonplaten) [inline .github/workflows/pr_tests.yml:36]:
```suggestion
          python -m pip install git+https://github.com/huggingface/accelerate
          python -m pip install -e .[quality,test]
```
This should work I think :-)

File: .github/workflows/push_tests.yml
Comment (patrickvonplaten) [inline .github/workflows/push_tests.yml:38]:
```suggestion
          python -m pip install git+https://github.com/huggingface/accelerate
          python -m pip install -e .[quality,test]
```

File: .github/workflows/pr_tests.yml
Comment (piEsposito) [reply .github/workflows/pr_tests.yml:36]: I tried putting the accelerate from master after the `quality,test` to ensure this is the version we will keep for the tests.

File: .github/workflows/push_tests.yml
Comment (piEsposito) [reply .github/workflows/push_tests.yml:38]: I tried putting the accelerate from master after the `quality,test` to ensure this is the version we will keep for the tests. How does that sound?

**Comment by patrickvonplaten** [comment]
Actually you seem to have figured out already how the githuh actions work :-) Think you'll just need to install accelerate before `pip install . [dev]` (maybe it also work the way you've done it now though) - in any way it's the right approach :-)

**Comment by piEsposito** [comment]
> Actually you seem to have figured out already how the githuh actions work :-) Think you'll just need to install accelerate before `pip install . [dev]` (maybe it also work the way you've done it now though) - in any way it's the right approach :-)

Actually, I think you were right. We can install it before because it will comply with `accelerate>=0.11.0`. Do you want me to change that?

**Comment by piEsposito** [comment]
@patrickvonplaten it worked, only the MPS tests on Apple M1 are not passing, but I think that's happening on the other PRs too right?

**Comment by piEsposito** [comment]
Now I think we are good to go.

**Comment by piEsposito** [comment]
@patrickvonplaten, can we add this funcionality to the other Stable Diffusion Pipelines on follow up PRs?

**Comment by piEsposito** [comment]
@patrickvonplaten, sorry to bother, can I do anything to help moving on with this PR?

**Review by patrickvonplaten — APPROVED** [review]
Great PR @piEsposito - sorry for replying so late! I was a bit swamped with issues :sweat_smile:

**Comment by patrickvonplaten** [comment]
@anton-l FYI regarding the tests -> until the next `accelerate` release we'll install from `main` to make these tests pass

**Comment by piEsposito** [comment]
@patrickvonplaten anything I can do to help with the issues? 

Thank you for merging that one. Can I do the same for the other stable diffusion pipelines?

**Comment by patrickvonplaten** [comment]
Yes I think adding this to the img2img and also inpaint pipeline makes a lot of sense :-) Also would you be interested in adding a section about this feature and the max memory usage to: https://huggingface.co/docs/diffusers/optimization/fp16 maybe?

**Comment by piEsposito** [comment]
All right, I will do that on the next few days. About the documentation: how can I do that?

**Comment by patrickvonplaten** [comment]
@piEsposito regarding the documentation you simply need to change the corresponding files here: https://github.com/huggingface/diffusers/tree/main/docs/source as well as add docstring to your method. 
BTW I've done some renaming here: https://github.com/huggingface/diffusers/pull/1016/files

**Comment by patrickvonplaten** [comment]
I think as soon as we have docs up as mentioned here: https://github.com/huggingface/diffusers/pull/850#issuecomment-1292320430 we can promote this super cool new reduction to less than 2GB RAM, what do you think? :-)

**Comment by piEsposito** [comment]
@patrickvonplaten great idea, I will try to do that till the end of the week. Should I add it to https://github.com/huggingface/diffusers/blob/main/docs/source/optimization/fp16.mdx ?

If we join this with attention slicing, it gets as small as 700mb.

**Pull request merged**
